### PR TITLE
Handle seqence length == 0 for RNN

### PR DESCRIPTION
**Pull request by yufenglee**
Following the convention we use in ReverseSequence, we don't need to do calculation for sequence length 0.

### CODE REVIEW

File: onnxruntime/core/providers/cpu/rnn/deep_cpu_gru.cc [hunk 1]
```diff
@@ -561,6 +561,8 @@ void UniDirectionalGru<T>::Compute(const gsl::span<const T>& inputs_arg,
   int32_t min_sequence_length = std::min(seq_length_, *std::min_element(sequence_lengths.cbegin(),
                                                                         sequence_lengths.cend()));
 
+  if (max_sequence_length == 0) return;
+
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/rnn/deep_cpu_gru.cc]: This differs slightly in that we zero out the output for entries beyond the sequence length. So if max_sequence_length is 0 should we call fill_n to blank the output for consistency?

      for (int r = 0; r < batch_size_; r++) {
        if (step >= min_sequence_length && step >= sequence_lengths[r]) {
          if (output_sequence) {
            auto fill_output = output + r * hidden_size_;
            std::fill_n(fill_output, hidden_size_, T{});
          }

          continue;
        }

**Comment by skottmckay** [comment]
Unit tests would be good to validate a) when one length is 0, and b) when all lengths are zero.

File: onnxruntime/core/providers/cpu/rnn/deep_cpu_gru.cc (hunk 1 above)
Comment (yufenglee) [reply onnxruntime/core/providers/cpu/rnn/deep_cpu_gru.cc]: >   for (int r = 0; r < batch_size_; r++) {
    if (step >= min_sequence_length && step >= sequence_lengths[r]) {
      if (output_sequence) {
        auto fill_output = output + r * hidden_size_;
        std::fill_n(fill_output, hidden_size_, T{});
      }

      continue;
    }

This logic zero out the output value between [sequence_lengths[r], max_sequence_length). As the max_sequence_length is already 0. Just return is consistent with the logic. 

However, it's a good practice to zero out the memory and make the output determined. I will zero out for the whole output.

File: onnxruntime/core/providers/cpu/rnn/deep_cpu_gru.cc [hunk 2]
```diff
@@ -1024,20 +1034,24 @@ void UniDirectionalGru<T>::Compute(const gsl::span<const T>& inputs_arg,
...
 
-  if (output_sequence) {
-    // copy last output to final_hidden_state
-    for (int i = 0; i < batch_size_; i++) {
-      const int seq_len = sequence_lengths[i];
+  // copy last output to final_hidden_state
+  for (int i = 0; i < batch_size_; i++) {
+    const int seq_len = sequence_lengths[i];
+    if (seq_len == 0) {
+      auto final_hidden_state_dst = final_hidden_state.begin() + i * hidden_size_;
```
Comment (skottmckay) [inline onnxruntime/core/providers/cpu/rnn/deep_cpu_gru.cc]: We don't need to do the fill if output_sequence is false do we?

Comment (yufenglee) [reply onnxruntime/core/providers/cpu/rnn/deep_cpu_gru.cc]: this one is to fill the Y_h, not Y. I think we need to fill it to keep consistent.

**Pull request merged**